sklearn/utils/validation.py

```diff
@@ -2610,3 +2611,281 @@ def _to_object_array(sequence):
+
+        unexpected_names = sorted(X_feature_names_set - fitted_feature_names_set)
+        missing_names = sorted(fitted_feature_names_set - X_feature_names_set)
+
+        def add_names(names):
+            output = ""
+            max_n_names = 5
+            for i, name in enumerate(names):
+                if i >= max_n_names:
+                    output += "- ...\n"
+                    break
+                output += f"- {name}\n"
+            return output
+
+        if unexpected_names:
+            message += "Feature names unseen at fit time:\n"
+            message += add_names(unexpected_names)
+
+        if missing_names:
+            message += "Feature names seen at fit time, yet now missing:\n"
+            message += add_names(missing_names)
+
+        if not missing_names and not unexpected_names:
+            message += "Feature names must be in the same order as they were in fit.\n"
+
+        raise ValueError(message)
+
+
+def _check_n_features(estimator, X, reset):
+    """Set the `n_features_in_` attribute, or check against it on an estimator.
+
+    .. versionchanged:: 1.6
+        Moved from `BaseEstimator` to `utils.validation`.
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        The estimator to validate the input for.
+
+    X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+        The input samples.
+
+    reset : bool
+        If True, the `n_features_in_` attribute is set to `X.shape[1]`.
+        If False and the attribute exists, then check that it is equal to
+        `X.shape[1]`. If False and the attribute does *not* exist, then
+        the check is skipped.
+        .. note::
+           It is recommended to call reset=True in `fit` and in the first
+           call to `partial_fit`. All other methods that validate `X`
+           should set `reset=False`.
+    """
+    try:
+        n_features = _num_features(X)
+    except TypeError as e:
+        if not reset and hasattr(estimator, "n_features_in_"):
+            raise ValueError(
+                "X does not contain any features, but "
+                f"{estimator.__class__.__name__} is expecting "
+                f"{estimator.n_features_in_} features"
+            ) from e
+        # If the number of features is not defined and reset=True,
+        # then we skip this check
+        return
+
+    if reset:
+        estimator.n_features_in_ = n_features
+        return
+
+    if not hasattr(estimator, "n_features_in_"):
+        # Skip this check if the expected number of expected input features
+        # was not recorded by calling fit first. This is typically the case
+        # for stateless transformers.
+        return
+
+    if n_features != estimator.n_features_in_:
+        raise ValueError(
+            f"X has {n_features} features, but {estimator.__class__.__name__} "
+            f"is expecting {estimator.n_features_in_} features as input."
+        )
+
+
+def validate_data(
+    _estimator,
```

Comment:
I'm not sure about what the `_` convention means here.

Reply:
This is a convention taken from cpython. Since this argument is positional only, the user would never type `_estimator`.

Reply:
OK. So we never used this convention before, I don't know if we should for this particular function. But this is also the first time that we use `/`.

Reply:
We already have it:

https://github.com/scikit-learn/scikit-learn/blob/d8f07815e3d3b234889c86cd9575bf34f0e15cbe/sklearn/utils/_metadata_requests.py#L1522

Reply:
Actually, we use it in metadata routing so fine with me.